server/src/main/java/org/opensearch/cluster/coordination/DecommissionNodeAttributeClusterStateTaskExecutor.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.coordination;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.ClusterStateTaskExecutor;
+import org.opensearch.cluster.ClusterStateTaskListener;
+import org.opensearch.cluster.decommission.DecommissionAttribute;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.cluster.routing.allocation.AllocationService;
+import org.opensearch.persistent.PersistentTasksCustomMetadata;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * Decommissions and shuts down nodes having a given attribute and updates the cluster state
+ *
+ * @opensearch.internal
+ */
+public class DecommissionNodeAttributeClusterStateTaskExecutor
```

Comment:
should we create a `NodesRemovalClusterStateTaskExecutor` instead which can be reused ? This will tie it to be used only in case of decommissioning .

Reply:
Can you extend the existing ```NodeRemovalClusterStateTaskExecutor``` class itself to handle extra attributes for evaluation with overloaded method instead of having one more executor.
Also there is subtle diff between decommissioning and node removal. Decommissioned node should never join back but node which left the cluster and can join any time later. I hope this will be taken care during JoinValidator logic

Reply:
Today, there already exists `NodeRemovalClusterStateTaskExecutor` which takes care of removing a node from the cluster state. This particular executor takes care of removing nodes based on the attributes it has. If needed, we can remove the word `Decommission` so that it doesn't just bind with decommissioning

Reply:
> Also there is subtle diff between decommissioning and node removal. Decommissioned node should never join back but node which left the cluster and can join any time later. I hope this will be taken care during JoinValidator logic

Yes, join validator will have the logic to prevent decommissioned nodes from joining back

Reply:
> Today, there already exists `NodeRemovalClusterStateTaskExecutor` which takes care of removing a node from the cluster state. This particular executor takes care of removing nodes based on the attributes it has. If needed, we can remove the word `Decommission` so that it doesn't just bind with decommissioning

I was proposing that the `TransportDecommission` finds out the list of nodes from the attributes and then use this executor.   Hence it will be totally decoupled from `Decommission` .

Reply:
The problem with list and then remove is the partial state in between which can be tricky. We need to make sure decommission of an attribute like zone is atomic in nature